### DISCUSSION TITLE
Disable TLSv1 1.1 & SSL Along With Deprecated Ciphers

### DISCUSSION THREAD

**Discussion by IamNishN** [DOCGroup/ACE_TAO#1485] [Q&A]
Hello Team,

We are using ace 6.5.8 on Oracle Linux 6. We are using ace to create a socket and start server. Issue is that after i start my service running nmap command gives output that SSLv3, TLSv1 and 1.1 are in use while as per security guidelines only TLS 1.2 is allowed. I am trying to figure out how to disable all these versions.
We are using ACE_SSL_Context to pass certificate and keys. I tried using filter_versions function to filter TLS versions and set_mode but it is not working. Can you please help me how we can disable these outdated SSL versions and ciphers.

Thanks,
Nishchal

**Comment by ferchor2003** [comment]
Nishchal,
When you create your ACE_SSL_Context you can set the options to do it:
```
        // Do not use the SSLv2 protocol.As of OpenSSL 1.0.2g the SSL_OP_NO_SSLv2 option is set by default.
        SSL_CTX_set_options(ssl_context->context(), SSL_OP_NO_SSLv2);
        // Do not use the SSLv3 protocol. It is recommended that applications should set this option.
        SSL_CTX_set_options(ssl_context->context(), SSL_OP_NO_SSLv3);

        // ACE_SSL_Context::SSLv23 is used to negotiate the actual protocol version 
        // to the highest version mutually supported by the client and the server. 
        // The supported protocols are SSLv3, TLSv1, TLSv1.1 and TLSv1.2.
        // ACE 7.0.3 uses SSLv23_method(), SSLv23_server_method(), SSLv23_client_method() which are deprecated.
        // Should use TLS_method(), TLS_server_method(), TLS_client_method() instead.
        ssl_context->set_mode(ACE_SSL_Context::SSLv23);

```

**Reply by djanello** [reply]
This fixes the problem.
For benefit of others running into this, here are the exact error messages that get triggered.
 Client: 
ACE_SSL (2061729|2061731) error code: 336151568 - error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure

Server:
ACE_SSL (2061728|2061728) error code: 337092801 - error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher

Hopefully the search engines pick up these error messages, the top search results right now lead to dead ends related to SSL security levels etc that don't fix the problem.
